src/persist-client/src/shard_id.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::str::FromStr;
+
+use bytes::BufMut;
+use proptest_derive::Arbitrary;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use mz_persist_types::Codec;
+use mz_repr::proto::newapi::{RustType, TryFromProtoError};
+
+include!(concat!(env!("OUT_DIR"), "/mz_persist_client.shard_id.rs"));
+
+/// An opaque identifier for a persist durable TVC (aka shard).
+///
+/// The [std::string::ToString::to_string] format of this may be stored durably
+/// or otherwise used as an interchange format. It can be parsed back using
+/// [str::parse] or [std::str::FromStr::from_str].
+#[derive(Arbitrary, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct ShardId([u8; 16]);
+
+impl RustType<ProtoShardId> for ShardId {
+    fn into_proto(self: &Self) -> ProtoShardId {
+        ProtoShardId {
+            shard_id: self.to_string()
```

Comment:
Its a little weird that this and `Codec` use `to_string`, but they are not defined in terms of each other

Reply:
well, due to using `to_string` they are already sharing the `Display` implementation. I guess the `Codec` implementation could be entirely replaced in the future with just regular Protobuf serialization? I'm a bit less familiar with how we're using `Codec` right now.